File: py/microbench_sum_x.jl

include("microbench_util.jl")

function sum_x(positions)
    result = 0.
    for i in eachindex(positions)
        result += positions[i].x
    end
    return result
end

function get_x(vec, index)
    return vec[index].x
end

number_particles = 1024
nb_steps = 200

positions = Vector{Point3D}(undef, number_particles)

x = 0.0
for i in eachindex(positions)
    positions[i] = Point3D(x, 0, 0)
    global x += 1.0
end

print("sum_x(positions)")
@btime sum_x(positions)
print("get_x(positions, 2)")
@btime get_x(positions, 2)

File: py/microbench_sum_x.py
from vector import Vector

from microbench_pypy4 import Point3D, number_particles
from microbench_pypy_list import Point3D as P3D_list


def sum_x(positions):
    result = 0.0
    for i in range(len(positions)):
        result += positions[i].x
    return result


if __name__ == "__main__":

    positions = Vector[Point3D].zeros(number_particles)

    x = 0.0
    for position in positions:
        position.x = x
        x += 1.0

    positions_list = Vector[P3D_list].zeros(number_particles)

    x = 0.0
    for position in positions_list:
        position.x = x
        x += 1.0

    assert sum_x(positions_list) == sum_x(positions)


    from IPython import get_ipython
    ipython = get_ipython()

    print("sum_x(positions)")
    ipython.magic("timeit sum_x(positions)")
    print("sum_x(positions_list)")
    ipython.magic("timeit sum_x(positions_list)")
